File: 0287_find_duplicate_num/find_duplicate.py
from typing import List


def find_duplicate(nums: List[int]) -> int:
    """
    Time: O(n)
    Space: O(1)
    """
    h = t = nums[0]
    while True:
        t, h = nums[t], nums[nums[h]]
        if h == t:
            break
    h = nums[0]
    while h != t:
        t, h = nums[t], nums[h]
        if h == t:
            break
    return h

File: 0287_find_duplicate_num/find_duplicate.go
package main

/**
 * Time: O(n)
 * Space: O(1)
 */
func findDuplicate(nums []int) int {
	t, h := nums[0], nums[0]
	for {
		t = nums[t]
		h = nums[nums[h]]
		if t == h {
			break
		}
	}
	h = nums[0]
	for t != h {
		h = nums[h]
		t = nums[t]
	}
	return h
}

func main() {

}
